### PR TITLE
replaced  the autocorrelation calculation with a proper FFT based one

### PR DESCRIPTION
**Pull request by wdpozzo**

### CODE REVIEW

**Comment by wdpozzo** [comment]
Hi  John, this is long overdue. Note that, the proper calculation always  finds extremely low ACLs, order 20.

File: cpnest/sampler.py [hunk 1]
```diff
@@ -122,15 +127,20 @@ def reset(self):
...
             _, p = next(self.yield_sample(-np.inf))
-        if self.verbose >= 3:
+            self.estimate_nmcmc_on_the_fly()
+        
+        if self.sample_prior is True:
             # save the poolsize as prior samples
             
             prior_samples = []
-            for k in tqdm(range(self.maxmcmc), desc='SMPLR {} generating prior samples'.format(self.thread_id),
+            for k in tqdm(range(5000), desc='SMPLR {} generating prior samples'.format(self.thread_id),
```
Comment (johnveitch) [inline cpnest/sampler.py]: Why 5000? Probably better change this back as it is unrelated to the ACL?

File: cpnest/nest2pos.py [hunk 2]
```diff
@@ -171,14 +171,17 @@ def autocorrelation(x):
...
+    sf=cf.conjugate()*cf
+    corr=np.fft.ifft(sf).real/var/len(x)
+
+    return corr
+
+def acl(acf, safety = 5.0):
     for i in range(len(acf)):
-        if acf[i] < i/5.0:
-            return acf[i]+1
+        if np.abs(acf[i]) < i/safety:
```
Comment (johnveitch) [inline cpnest/nest2pos.py]: The [autocorrelation length used in lalinference](https://git.ligo.org/lscsoft/lalsuite/-/blob/master/lalinference/lib/LALInferenceNestedSampler.c#L1261) is defined differently.
Let 
L = \sum_i^j 2*Cxx[i]
Where Cxx[i] is the autocorrelation at lag i for parameter x.
Then the autocorrelation length is the value of L when Cxx[j]<TOLERANCE, where TOLERANCE is 0.01 in LALInferenceNest.

**Review by johnveitch — CHANGES_REQUESTED** [review]
I don't think the method implemented is right. Take a look at https://arxiv.org/abs/1011.0175 for various options - they found that the AR-process method is most robust but I haven't implemented it.
The implementation of the LALInference version is given by:
```
def acl(x, tolerance=0.01):
    """
    Compute autocorrelation time for x
    """
    T=1
    i=1
    Cxx=1
    y=x-x.mean()
    while Cxx>tolerance and i<len(y):
        Cxx = scipy.stats.pearsonr(y[:-i],y[i:])[0]
        T+=2*Cxx
        i+=1
    return T
```

**Comment by wdpozzo** [comment]
Ok, I implemented the changes you suggested. I did not use pearsonr because is extremely slow and gives identical results to just computing the whole acf with an FFT. Also, i implemented the AR method, that is commented out. it is unfeasibly slow, so it can be removed if  you want to.

**Comment by johnveitch** [comment]
Thanks Walter. I made a few minor tweaks. The HMC 2d test is failing with
```
john@athena:~/src/cpnest$ python3 tests/test_2d_hmc.py
Checkpoint exception raise
test_run (__main__.GaussianTestCase) ... 2020-05-28, 10:06:34 - CPNest  : Setting verbosity to 1
2020-05-28, 10:06:34 - CPNest  : Running with 1 parallel threads
Process Process-2:
Traceback (most recent call last):                                                                           | 0/1000 [00:00<?, ?it/s]
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/john/src/cpnest/cpnest/sampler.py", line 202, in produce_sample
    self._produce_sample()
  File "/home/john/src/cpnest/cpnest/sampler.py", line 214, in _produce_sample
    self.reset()
  File "/home/john/src/cpnest/cpnest/sampler.py", line 138, in reset
    _, p = next(self.yield_sample(-np.inf))
  File "/home/john/src/cpnest/cpnest/sampler.py", line 393, in yield_sample
    p.update_trajectory_length(self.estimate_nmcmc())
  File "/home/john/src/cpnest/cpnest/sampler.py", line 190, in estimate_nmcmc
    ACL = [acl(samples[:,i]) for i in range(samples.shape[1])]
IndexError: tuple index out of range
```

File: cpnest/sampler.py [hunk 3]
```diff
@@ -167,7 +177,27 @@ def estimate_nmcmc(self, safety=5, tau=None):
...
+        # the stored samples
+        self.ACL = []
+        samples = np.array(self.samples)
 
+        if self.store_chain is False:
+            self.ACL = [acl(samples[:,i]) for i in range(samples.shape[1])]
+        else:
+            self.ACL = [acl(samples[-self.maxmcmc:,i]) for i in range(samples.shape[1])]
+        self.Nmcmc = int(np.max(self.ACL))
+        if self.store_chain is False and len(self.samples) > self.maxmcmc:
```
Comment (johnveitch) [inline cpnest/sampler.py]: I don't understand the logic behind this. Can you add a comment?

Comment (wdpozzo) [reply cpnest/sampler.py]: done.  the idea is that if you are storing the full mcmc chain, then you should use  only the last maxmcmc samples to compute the ACL and not the full story

**Comment by johnveitch** [comment]
Thanks Walter. I am still seeing a problem with the HMC test, it's failing a bit later though. Do you see the same thing?

**Comment by wdpozzo** [comment]
The last commit should fix it.

**Comment by johnveitch** [comment]
Thanks. I am seeing an error when running with verbose=3 (in the examples/eggbox.py program). It looks like it's unable to stack prior samples when preparing to plot:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
~/src/cpnest/examples/eggbox.py in <module>
     39     M = EggboxModel()
     40     work=cpnest.CPNest(M,verbose=3,nthreads=4,nlive=1000,maxmcmc=1000,poolsize=1000)
---> 41     work.run()
     42

~/src/cpnest/cpnest/cpnest.py in run(self)
    224             sys.exit(130)
    225
--> 226         self.posterior_samples = self.get_posterior_samples(filename=None)
    227         if self.verbose>1: self.plot(corner = False)
    228

~/src/cpnest/cpnest/cpnest.py in get_posterior_samples(self, filename)
    292
    293             # first deal with the prior samples
--> 294             self.prior_samples = stack_arrays([p for p in prior_samples])
    295             if filename:
    296                 np.savetxt(os.path.join(

<__array_function__ internals> in stack_arrays(*args, **kwargs)

/usr/lib/python3/dist-packages/numpy/lib/recfunctions.py in stack_arrays(arrays, defaults, usemask, asrecarray, autoconvert)
   1300     fldnames = [d.names for d in ndtype]
   1301     #
-> 1302     dtype_l = ndtype[0]
   1303     newdescr = _get_fieldspec(dtype_l)
   1304     names = [n for n, d in newdescr]

IndexError: list index out of range
```

I know it's not caused by these changes but is there any chance you could take a look?

**Comment by wdpozzo** [comment]
wow that is a big job! the code is full  of inconsistencies  for prior_sampling and verbose. will try and fix it in the next few hours

**Pull request merged**